sdk/servicebus/azure-servicebus/azure/servicebus/_common/message.py

```diff
@@ -558,11 +564,16 @@ def dead_letter(self, reason=None, description=None):
             MGMT_REQUEST_DEAD_LETTER_REASON: str(reason) if reason else "",
             MGMT_REQUEST_DEAD_LETTER_DESCRIPTION: str(description) if description else ""}
         try:
-            self._receiver._settle_message(
-                SETTLEMENT_DEADLETTER,
-                [self.lock_token],
-                dead_letter_details=details
-            )
+            if self._is_deferred_message:
+                self._receiver._settle_message(
+                    SETTLEMENT_DEADLETTER,
+                    [self.lock_token],
+                    dead_letter_details=details
+                )
+            else:
+                # note: message.reject() can not set reason and description properly due to the issue
+                # https://github.com/Azure/azure-uamqp-python/issues/155
+                self.message.reject(condition=DEADLETTERNAME, description=description)
```

Comment:
Weren't we going to have functionality wherein if this failed (if the link is closed or was replaced with a new link, thus disallowing settlement) we'd fall back to using the mgmt. call?

Reply:
code updated, thanks for the callout